### PR TITLE
payload type consistency fixes

### PR DESCRIPTION
**Pull request by caneleex**
this PR fixes types of some payload fields to match the behavior of the Discord API.

### CODE REVIEW

File: discord/thread.go [hunk 1]
```diff
@@ -3,15 +3,16 @@ package discord
 import "github.com/disgoorg/disgo/json"
 
 type ThreadCreateFromMessage struct {
-	Name                string              `json:"name"`
-	AutoArchiveDuration AutoArchiveDuration `json:"auto_archive_duration,omitempty"`
+	Name                string               `json:"name"`
+	AutoArchiveDuration *AutoArchiveDuration `json:"auto_archive_duration,omitempty"`
```
Comment (TopiSenpai) [inline discord/thread.go:7]: `0` gets omitted

File: discord/thread.go [hunk 2]
```diff
@@ -3,15 +3,16 @@ package discord
...
+	RateLimitPerUser    *int                 `json:"rate_limit_per_user,omitempty"`
 }
 
 type ForumThreadCreate struct {
-	Name                string              `json:"name"`
-	AutoArchiveDuration AutoArchiveDuration `json:"auto_archive_duration,omitempty"`
-	RateLimitPerUser    int                 `json:"rate_limit_per_user,omitempty"`
-	Message             MessageCreate       `json:"message"`
+	Name                string               `json:"name"`
+	AutoArchiveDuration *AutoArchiveDuration `json:"auto_archive_duration,omitempty"`
```
Comment (TopiSenpai) [inline discord/thread.go:13]: `0` gets omitted

File: discord/thread.go [hunk 3]
```diff
@@ -3,15 +3,16 @@ package discord
 import "github.com/disgoorg/disgo/json"
 
 type ThreadCreateFromMessage struct {
-	Name                string              `json:"name"`
-	AutoArchiveDuration AutoArchiveDuration `json:"auto_archive_duration,omitempty"`
+	Name                string               `json:"name"`
+	AutoArchiveDuration *AutoArchiveDuration `json:"auto_archive_duration,omitempty"`
+	RateLimitPerUser    *int                 `json:"rate_limit_per_user,omitempty"`
```
Comment (TopiSenpai) [inline discord/thread.go:8]: `0` is the default value

File: discord/thread.go [hunk 4]
```diff
@@ -3,15 +3,16 @@ package discord
...
 }
 
 type ForumThreadCreate struct {
-	Name                string              `json:"name"`
-	AutoArchiveDuration AutoArchiveDuration `json:"auto_archive_duration,omitempty"`
-	RateLimitPerUser    int                 `json:"rate_limit_per_user,omitempty"`
-	Message             MessageCreate       `json:"message"`
+	Name                string               `json:"name"`
+	AutoArchiveDuration *AutoArchiveDuration `json:"auto_archive_duration,omitempty"`
+	RateLimitPerUser    *int                 `json:"rate_limit_per_user,omitempty"`
```
Comment (TopiSenpai) [inline discord/thread.go:14]: `0` is the default value

File: discord/thread.go [hunk 5]
```diff
@@ -33,8 +34,8 @@ type ThreadCreate interface {
 }
 
 type GuildNewsThreadCreate struct {
-	Name                string              `json:"name"`
-	AutoArchiveDuration AutoArchiveDuration `json:"auto_archive_duration,omitempty"`
+	Name                string               `json:"name"`
+	AutoArchiveDuration *AutoArchiveDuration `json:"auto_archive_duration,omitempty"`
```
Comment (TopiSenpai) [inline discord/thread.go:38]: `0` gets omitted

File: discord/thread.go [hunk 6]
```diff
@@ -53,8 +54,8 @@ func (GuildNewsThreadCreate) Type() ChannelType {
 }
 
 type GuildPublicThreadCreate struct {
-	Name                string              `json:"name"`
-	AutoArchiveDuration AutoArchiveDuration `json:"auto_archive_duration,omitempty"`
+	Name                string               `json:"name"`
+	AutoArchiveDuration *AutoArchiveDuration `json:"auto_archive_duration,omitempty"`
```
Comment (TopiSenpai) [inline discord/thread.go:58]: `0` gets omitted

File: discord/thread.go [hunk 7]
```diff
@@ -73,9 +74,9 @@ func (GuildPublicThreadCreate) Type() ChannelType {
 }
 
 type GuildPrivateThreadCreate struct {
-	Name                string              `json:"name"`
-	AutoArchiveDuration AutoArchiveDuration `json:"auto_archive_duration,omitempty"`
-	Invitable           bool                `json:"invitable"`
+	Name                string               `json:"name"`
+	AutoArchiveDuration *AutoArchiveDuration `json:"auto_archive_duration,omitempty"`
```
Comment (TopiSenpai) [inline discord/thread.go:78]: `0` gets omitted

File: discord/thread.go [hunk 8]
```diff
@@ -73,9 +74,9 @@ func (GuildPublicThreadCreate) Type() ChannelType {
...
+	Invitable           *bool                `json:"invitable"`
```
Comment (TopiSenpai) [inline discord/thread.go:79]: does this default to false? we should check

File: discord/user.go [hunk 9]
```diff
@@ -127,6 +127,6 @@ const (
 
 // SelfUserUpdate is the payload used to update the OAuth2User
 type SelfUserUpdate struct {
-	Username string               `json:"username"`
-	Avatar   *json.Nullable[Icon] `json:"avatar"`
+	Username *string              `json:"username,omitempty"`
```
Comment (TopiSenpai) [inline discord/user.go:130]: username can't be empty so this does not need to be a pointer

File: discord/voice_state.go [hunk 10]
```diff
@@ -23,8 +23,12 @@ type VoiceState struct {
 	RequestToSpeakTimestamp *time.Time    `json:"request_to_speak_timestamp"`
 }
 
-type UserVoiceStateUpdate struct {
-	ChannelID               snowflake.ID              `json:"channel_id"`
-	Suppress                *bool                     `json:"suppress,omitempty"`
+type CurrentUserVoiceStateUpdate struct {
+	UserVoiceStateUpdate
```
Comment (TopiSenpai) [inline discord/voice_state.go:27]: can we just paste the 2 fields? as making a struct like this is kinda annoying

File: discord/voice_state.go [hunk 11]
```diff
@@ -23,8 +23,12 @@ type VoiceState struct {
...
 	RequestToSpeakTimestamp *json.Nullable[time.Time] `json:"request_to_speak_timestamp,omitempty"`
 }
+
+type UserVoiceStateUpdate struct {
+	ChannelID *json.Nullable[snowflake.ID] `json:"channel_id,omitempty"`
```
Comment (TopiSenpai) [inline discord/voice_state.go:32]: `0` should kick the user out of a channel

**Review by TopiSenpai — CHANGES_REQUESTED** [review]

File: discord/role.go [hunk 12]
```diff
@@ -53,30 +53,30 @@ type RoleTag struct {
...
-	Mentionable *bool                `json:"mentionable"`
+	Mentionable *bool                `json:"mentionable,omitempty"`
 }
 
 // RolePositionUpdate is the payload to update a Role(s) position
 type RolePositionUpdate struct {
-	ID       snowflake.ID `json:"id"`
-	Position *int         `json:"position"`
+	ID       snowflake.ID        `json:"id"`
+	Position *json.Nullable[int] `json:"position,omitempty"`
```
[the remark this answers is not in the document]
Comment (caneleex) [reply discord/role.go:79]: it is optional. it just resets the position and moves the role all the way to the bottom.

[the remark this answers is not in the document]
Comment (TopiSenpai) [reply discord/role.go:79]: the bottom would be the same as 0 I guess?

File: discord/guild_scheduled_event.go [hunk 13]
```diff
@@ -35,21 +36,22 @@ type GuildScheduledEventCreate struct {
...
-	EntityMetaData     *EntityMetaData             `json:"entity_metadata,omitempty"`
-	Name               string                      `json:"name,omitempty"`
-	PrivacyLevel       *ScheduledEventPrivacyLevel `json:"privacy_level,omitempty"`
-	ScheduledStartTime *time.Time                  `json:"scheduled_start_time,omitempty"`
-	ScheduledEndTime   *time.Time                  `json:"scheduled_end_time,omitempty"`
-	Description        *string                     `json:"description,omitempty"`
-	EntityType         *ScheduledEventEntityType   `json:"entity_type,omitempty"`
-	Status             *ScheduledEventStatus       `json:"status,omitempty"`
+	ChannelID          *snowflake.ID                  `json:"channel_id,omitempty"`
+	EntityMetaData     *json.Nullable[EntityMetaData] `json:"entity_metadata,omitempty"`
```
Comment (TopiSenpai) [inline discord/guild_scheduled_event.go:47]: I don't think this needs to be a `json.Nullable[T]` as the fields inside are the important part here. You can reset the fields inside by making them empty

**Review by TopiSenpai — CHANGES_REQUESTED** [review]
some minor stuff I noticed

**Review by TopiSenpai — APPROVED** [review]

**Pull request merged**